content: source track: update source threats for draft spec (#1236)
slsa-framework/slsa: docs/spec/draft/threats.md

A source integrity threat is a potential for an adversary to introduce a change
to the source code that does not reflect the intent of the software producer.
This includes modification of the source data at rest as well as insider threats,
when an authorized individual introduces an unauthorized change.

### (A) Producer

The producer of the software intentionally produces code that harms the
consumer, or the producer otherwise uses practices that are not deserving of the
consumer's trust.

<details><summary>Software producer intentionally creates a malicious revision of the source</summary>

*Threat:* An producer intentionally creates a malicious revision with the intent of harming their consumers.

*Mitigation:*
This kind of attack cannot be directly mitigated through SLSA controls.
Tools like the [OSSF Scorecard](https://github.com/ossf/scorecard) can help to quantify the risk of consuming artifacts from specific organizations, but do not fully remove it.

> Trustworthiness scales with transparency, and consumers SHOULD push their vendors to follow transparency best-practices.

Reviewer: "Transparency" is unfortunately an overloaded word in our world. WDYM specifically in this context?

Author: oof, that's a good callout. 

I propose that I mean: 
* open source
* open build
* open policies
* published "definitions of correctness."

Author: @TomHennen I think SLSA tooling gives you a way to unilaterally improve trust when the source and builds are open source. This is possible independently of whether or not the producer has something to lose.

Author: I left two suggestions. wdyt @trishankatdatadog @TomHennen ?

Reviewer: Left a comment down there...

Author: good ideas. I'll incorporate these.

When transparency is not possible, consumers may choose not to consume the artifact, or may require additional evidence of correctness from a trusted third-party.

> *Example:* The [libxz attack](https://access.redhat.com/security/cve/CVE-2024-3094) and the intentional sabotage of [colors.js and faker.js](https://fossa.com/blog/npm-packages-colors-faker-corrupted/) are clear examples of an organization with an otherwise good reputation intentionally producing malicious code.

Author: @TomHennen I think we missed the suggestion here ☝️

> *Example:* The [libxz attack](https://access.redhat.com/security/cve/CVE-2024-3094) and the intentional sabotage of [colors.js and faker.js](https://fossa.com/blog/npm-packages-colors-faker-corrupted/) are clear examples of an organization with an otherwise good reputation intentionally producing malicious code.

Author: ```suggestion
*Example:* A producer with an otherwise good reputation decides suddenly to produce a malicious artifact with the intent to harm their consumers. 
```
This is simpler. wdyt @TomHennen ?

</details>

### (B) Modifying the source

An adversary without any special administrator privileges attempts to introduce a change counter to the declared intent of the source by following the producer's official source control process.

Threats in this category can be mitigated by following source control management best practices.

#### (B1) Submit change without review

<details><summary>Directly submit without review</summary>

*Threat:* Submit code to the source repository without another person reviewing.

> *Mitigation:* The producer can require pre-approval for all changes.

Reviewer: This new language is actually more confusing than the last one. What's a "pre-approval"?

Author: ```suggestion
*Mitigation:* The producer requires approval of all changes before they are accepted.
```

Author: is that a clearer? pre-approval does sound like you're applying for a car loan lol

*Example:* Adversary directly pushes a change to a git repo's `main` branch.

> Solution: The producer can configure branch protection rules on the `main` branch.

Author: I think in many ways this section is duplicative of other sections. 

We could just say: 
```
*Threat*: source threats!
*Mitigation*: SLSA source track! <url>
```

> Solution: The producer can configure branch protection rules on the `main` branch.

Reviewer 2: Note that this page discusses lots of threats to the build and doesn't just defer to the SLSA Build Track.

The general purpose of this page, AIUI, was to centralize all the various threats in one place and direct folks elsewhere as needed.

> Solution: The producer can configure branch protection rules on the `main` branch.

Author: Potentially then most of these examples should link to relevant sections of the relevant tracks. Maybe we'll take it as a todo.

A best practice would be to required pre-approval of any changes via a change management tool (such a as GitHub pull request).

</details>
<details><summary>Single actor controls multiple accounts</summary>

*Threat:* An actor is able to control multiple account and effectively approve their own code changes.

> *Mitigation:* The produce must ensure that no actor is able to control or influence multiple accounts with review privileges.

Reviewer: ```suggestion
*Mitigation:* The producer must ensure that no actor is able to control or influence multiple accounts with review privileges.
```

*Example:* Adversary creates a pull request using a secondary account and approves it using their primary account.

Solution: The producer must require strongly authenticated user accounts and ensure that all accounts map to unique persons.
A common vector for this attack is to take over a robot account with the permission to contribute code.

> Control of the robot and an actors own legitimate account is enough to exploit this vulnerability.

Reviewer: Confusing. The solution goes into another problem, but does not prescribe a solution for that. Best to separate the latter out IMHO.

Author: I think we can just cut it. This is very duplicative of other sections in the source track.

A source integrity threat is a potential for an adversary to introduce a change
to the source code that does not reflect the intent of the software producer.
This includes modification of the source data at rest as well as insider threats,
when an authorized individual introduces an unauthorized change.

### (A) Producer

The producer of the software intentionally produces code that harms the
consumer, or the producer otherwise uses practices that are not deserving of the
consumer's trust.

<details><summary>Software producer intentionally creates a malicious revision of the source</summary>

> *Threat:* A producer intentionally creates a malicious revision with the intent of harming their consumers.

Author: @TomHennen what if we update this to cover both: 

```suggestion
*Threat:* A producer intentionally creates a malicious revision (or a VSA issuer intentionally creates a malicious attestation) with the intent of harming their consumers.
```

The mitigation for both is a clearer -- you cannot.

*Mitigation:*
This kind of attack cannot be directly mitigated through SLSA controls.

> Trustworthiness scales with transparency, and consumers SHOULD push their vendors to follow transparency best-practices.
> When transparency is not possible, consumers may choose not to consume the artifact, or may require additional evidence of correctness from a trusted third-party.
> Tools like the [OSSF Scorecard](https://github.com/ossf/scorecard) can help to quantify the risk of consuming artifacts from specific producers, but do not fully remove it.
> For example, a consumer may choose to only consume source artifacts from repositories that have a high score on the OSSF Scorecard.

Author: ```suggestion
Ultimately, consumers must decide which producers are trusted to produce artifacts and which issuers are trusted to produce VSAs. 
```

We could redirect from here over to "verifying platforms".

Reviewer: Sorry for late reply. Why do consumers even need to know which producers are trusted to produce artifacts (source code, I imagine, in this case)? Couldn't they just get away with the Source Track VSA from the issuer, and call it a day so long as its subjects/outputs (e.g., source code) matches the inputs to the next step (e.g., build)?

Tools like the [OSSF Scorecard](https://github.com/ossf/scorecard) can help to quantify the risk of consuming artifacts from specific producers, but do not fully remove it.
For example, a consumer may choose to only consume source artifacts from repositories that have a high score on the OSSF Scorecard.

*Threat:* Software producer intentionally submits "bad" code, following all
proper processes.

*Mitigation:* **TODO**

*Example:* A popular extension author sells the rights to a new owner, who then
modifies the code to secretly mine cryptocurrency at the users' expense. SLSA
does not protect against this, though if the extension were open source, regular
auditing may discourage this from happening.
*Example:* The [libxz attack](https://access.redhat.com/security/cve/CVE-2024-3094) and the intentional sabotage of [colors.js and faker.js](https://fossa.com/blog/npm-packages-colors-faker-corrupted/) are clear examples of an organization with an otherwise good reputation intentionally producing malicious code.

</details>

<!--
**TODO:** More producer threats? Perhaps the attack to xz where a malicious
contributor gained enhanced privileges through social engineering?

> -->

Author: we could add another section here like 
```
*threat*: A member of the producer's organization acts against the intent of the producer.
```
We could say this is when "the producer UNintentially ships bad code."

> A best practice would be to required pre-approval of any changes via a change management tool (such a as GitHub pull request).

Author: ```suggestion
A best practice is to require approval of any changes via a change management tool before they are accepted into the source.
```

> Solution: The producer must require strongly authenticated user accounts and ensure that all accounts map to unique persons.
> A common vector for this attack is to take over a robot account with the permission to contribute code.
> Control of the robot and an actors own legitimate account is enough to exploit this vulnerability.

Author: ```suggestion
Solution: The producer must require strongly authenticated user accounts and ensure that all accounts map to unique persons. It must not be possible for one actor to exploit robot actors that possess review privileges. 
```

A source integrity threat is a potential for an adversary to introduce a change
to the source code that does not reflect the intent of the software producer.
This includes the threat of an authorized individual introducing an unauthorized
change---in other words, an insider threat.
This includes modification of the source data at rest as well as insider threats,
when an authorized individual introduces an unauthorized change.

SLSA v1.0 does not address source threats, but we anticipate doing so in a

> [future version](current-activities.md#source-track). In the meantime, the

Author: I think this was a broken link? I have added a section to future-directions to target.

Reviewer 2: agreed

A source integrity threat is a potential for an adversary to introduce a change
to the source code that does not reflect the intent of the software producer.
This includes modification of the source data at rest as well as insider threats,
when an authorized individual introduces an unauthorized change.

SLSA does not yet address source threats, but we anticipate doing so in a
[future version](future-directions#source-track).

### (A) Producer

The producer of the software intentionally produces code that harms the
consumer, or the producer otherwise uses practices that are not deserving of the
consumer's trust.

<details><summary>Software producer intentionally creates a malicious revision of the source</summary>

> *Threat:* A producer intentionally creates a malicious revision (or a VSA issuer intentionally creates a malicious attestation) with the intent of harming their consumers.

Reviewer 2: ```suggestion
*Threat:* A producer intentionally creates a malicious revision with the intent of harming their consumers.
```

I think the VSA Issuer note should probably go in https://slsa.dev/spec/draft/threats#verification-threats (below) as it applies everywhere, and the 'producer' of the VSA isn't aligned with the producer being discussed here: https://slsa.dev/spec/draft/terminology#:~:text=Examples-,Producer,-A%20party%20who

*Mitigation:*
This kind of attack cannot be directly mitigated through SLSA controls.
Consumers must establish some basis to trust the organizations from which they consume software.
That basis may be:

> -   The code is open source and has a sufficiently large user-base that malicious changes are likely to be detected.

Author: I guess both to some extent. 
Expert code review may be "required" to approve something 
into the main branch, but prs are often reviewed by non-authorized (but maybe informed) reviewers too. 
Those extra eyes help.

Maybe this should be a time-to-mitigation claim. 
`With more sets of eyes, the length of time that a vulnerability exists in the repo is reduced.`

> -   The code is open source and has a sufficiently large user-base that malicious changes are likely to be detected.

Author: or maybe 

```The repo is open source with an active user-base. High numbers of engaged users increase the likelihood that bad code is detected during code review and reduce the time-to-detection when bad code makes it in.

> -   The code is open source and has a sufficiently large user-base that malicious changes are likely to be detected.

Reviewer 3: I think commenting about a reduced time to detection feels right. We can also make an additional comment about an "expert" review as a potential additional mitigation.

How about 
```suggestion
-   The code is open source and has a large engaged user-base. Time to detect and remediate malicious changes to the code base are reduced.
-   The organization requires that all code is vetted by expert reviewers before revisions are accepted.
```

-   The organization has sufficient legal or reputational incentives to dissuade it from making malicious changes.

Ultimately this is a judgement call with no straightforward answer.

*Example:* A producer with an otherwise good reputation decides suddenly to produce a malicious artifact with the intent to harm their consumers.

</details>

### (B) Modifying the source

An adversary without any special administrator privileges attempts to introduce a change counter to the declared intent of the source by following the producer's official source control process.

Threats in this category can be mitigated by following source control management best practices.

#### (B1) Submit change without review

<details><summary>Directly submit without review</summary>

> *Threat:* Submit code to the source repository without another person reviewing.

Reviewer 2: I think this leaves out that the code being submitted is malicious.

the vast majority of unreviewed changes aren't malicious.

Perhaps this can be rephrased as 

```suggestion
*Threat:* Malicious code submitted to the source repository.
```
And then the (partial) mitigation is "The producer requires approval of all changes before they are accepted".

Author: That's fine with me too. 
Generally, it's hard to say whether a commit is malicious, and "exploitable" ends up being just as bad.

*Mitigation:* The producer requires approval of all changes before they are accepted.

*Example:* Adversary directly pushes a change to a git repo's `main` branch.
Solution: The producer can configure branch protection rules on the `main` branch.
A best practice is to require approval of any changes via a change management tool before they are accepted into the source.

</details>
<details><summary>Single actor controls multiple accounts</summary>

*Threat:* An actor is able to control multiple account and effectively approve their own code changes.

*Mitigation:* The producer must ensure that no actor is able to control or influence multiple accounts with review privileges.

*Example:* Adversary creates a pull request using a secondary account and approves it using their primary account.

> Solution: The producer must require all accounts with 'write' and 'approval' permissions to be strongly authenticated and ensure they map to unique persons.

Reviewer 4: If two accounts under consideration must map to unique persons, don't we have also have the case that each person is mapped to at most one account?

That said, I think it is better to put the emphasis on the persons involved being distinct rather than the accounts, it would partially address the concern I had in the other thread I think.

> Solution: The producer must require all accounts with 'write' and 'approval' permissions to be strongly authenticated and ensure they map to unique persons.

Reviewer 2: I think the trick is that it should still be possible for anonymous folks to send PRs.  _They_ don't need to be unique.  But you'd want to unique and trusted people to be involved.

I also wonder how much of this nuance is actually needed.  These are just example solutions, they're not really meant to be completely bulletproof (I don't think?).

Another alternative is we just punt and say "SLSA doesn't address this yet".  But that seems annoying. :p

> Solution: The producer must require all accounts with 'write' and 'approval' permissions to be strongly authenticated and ensure they map to unique persons.

Reviewer 3: If an anonymous person sends a PR then this can be protected against by _requiring_ 2-party reviews. I don't know how to word that though.

You are probably right that this much nuance isn't needed here. This might even be addressing a separate threat?

> Solution: The producer must require all accounts with 'write' and 'approval' permissions to be strongly authenticated and ensure they map to unique persons.

Author: > An account that creates a revision request doesn't necessarily need to have write permissions. 

Not to get tooooooo in the weeds, but in most implementations an open requires write permissions to created the proposed merge commits for CI.

> Solution: The producer must require all accounts with 'write' and 'approval' permissions to be strongly authenticated and ensure they map to unique persons.

Author: I'm not sure how to tighten this one up. 
I feel like the threat, mitigation and example are all good, but the solution needs work.

> Solution: The producer must require all accounts with 'write' and 'approval' permissions to be strongly authenticated and ensure they map to unique persons.

Author: I could just adjust it -- "the producer much prevent this kind of thing from happening somehow"

> Solution: The producer must require strongly authenticated user accounts and ensure that all accounts map to unique persons.

Reviewer 2: Trying to scope this down some for folks.  There's no way an org can ensure all GitHub accounts map to unique persons.  They really just need it to be the case for people that can approve code.

```suggestion
Solution: The producer must require all accounts with 'write' and 'approval' permissions to be strongly authenticated and ensure they map to unique persons.
```

> A common vector for this attack is to take over a robot account with the permission to contribute code.

Author: yeah, I don't like the word mapping here. let me try to remove it.

> A common vector for this attack is to take over a robot account with the permission to contribute code.

Reviewer 2: optional: Robot accounts are discussed below so this seems somewhat redundant.  Suggest removing.

Author: That's probably fine. I _think_ these are different threat models, but not sure if the difference is worth calling out, or if these two blurbs are doing so. 
This one is about gaining access to two accounts with `write` and the robots one is about gaining access to an account with `bypass`. 

The mitigation for the first is _hard_ to execute and protect against. You'd have to somehow prevent account takeovers. 

The mitigation for the robots one is easier to execute (just run some kind of automation with a poisoned payload, usually) and protect against-- just do not allow any account to bypass code review requirements.

Reviewer 2: Right.  I guess I'll leave it to you?

A source integrity threat is a potential for an adversary to introduce a change
to the source code that does not reflect the intent of the software producer.
This includes modification of the source data at rest as well as insider threats,
when an authorized individual introduces an unauthorized change.

SLSA does not yet address source threats, but we anticipate doing so in a
[future version](future-directions#source-track).

### (A) Producer

The producer of the software intentionally produces code that harms the
consumer, or the producer otherwise uses practices that are not deserving of the
consumer's trust.

<details><summary>Software producer intentionally creates a malicious revision of the source</summary>

*Threat:* A producer intentionally creates a malicious revision with the intent of harming their consumers.

*Mitigation:*
This kind of attack cannot be directly mitigated through SLSA controls.
Consumers must establish some basis to trust the organizations from which they consume software.
That basis may be:

> -   The repo is open source with an active user-base. High numbers of engaged users may increase the likelihood that bad code is detected during code review and reduce the time-to-detection when bad code makes it in.

Reviewer 3: nit:
```suggestion
-   The repo is open source with an active user-base. High numbers of engaged users may increase the likelihood that bad code is detected during code review and reduce the time-to-detection when bad revisions are accepted.
```

or (B)?
-->

> <details><summary>The organization intentionally creates a malicious revision</summary>

Reviewer 4: Is there a reason to switch the terminology to "organization" from "producer"? Organization seems narrower to me, and the figure also uses producer.

Author: that's a good question. 
My idea was to align with the "entity that determines the intent of the source," because "sudden change of intent" is really the attack vector here.

Author: I started a conversation in slack. Copied here: 

https://openssf.slack.com/archives/C03NUSAPKC6/p1732542950838849?thread_ts=1732542830.795929&cid=C03NUSAPKC6

The threats diagram uses the term "producer" in the source threats section, but is not one of the [defined terms for the proposed source track](https://github.com/slsa-framework/slsa/blob/4a5c7e682c304f73a9d2501d3ec0c819648c346f/docs/spec/draft/source-requirements.md#definitions)

At least to me, "producer" doesn't capture what we mean here.
"producer" seems like it must be akin to "organization" in our source track terminology, but there are other actor-based threats against the source that do not require a compromised organization (IE: "contributors" and "reviewers" also "produce" source.) 

The source track needs to talk about several kinds of actors with roles [described here](https://github.com/slsa-framework/slsa/blob/4a5c7e682c304f73a9d2501d3ec0c819648c346f/docs/spec/draft/source-requirements.md#source-roles). The only entity that could be described as the sole "producer" would be the org, which defines the intent of the software and the rules to enforce it.

Reviewer 3: A producer can be an individual or an organization. I think it makes sense to keep it general.

> <details><summary>The organization changes the intent of the repository and creates a malicious revision</summary>

Reviewer 2: What about this...

```suggestion
<details><summary>The organization intentionally creates a malicious revision</summary>
```

Rationale: the "changes the intent of the repository" may be a distraction.  Folks might think "this thing still does compression so the major intent of the repo is the same"

*Example:* Adversary creates a pull request using a secondary account and approves it using their primary account.

> Solution: The producer must require all accounts with 'write' and 'approval' permissions to be strongly authenticated and ensure they map to unique persons.

Reviewer 4: ```suggestion
Solution: The producer must require all users trusted to write to the repository and approve changes to be strongly authenticated and ensure they map to unique persons.
```

I took a stab at generalizing this, WDYT?

Reviewer 2: To be honest, I think 'credentials' obscures the meaning.

'credentials' can be interpreted as a specific thing (e.g. password, cryptographic key, badge, etc).  I have _many_ of those things, and I use multiple of them when interacting with systems like GitHub.  Further those systems rarely restrict use based on the way a user authenticated.

Reviewer 4: That's a good point. Maybe:

```
The producer must require anyone trusted to write to the repository and approve changes to be strongly authenticated and ensure the credentials they use to authenticate themselves map to unique persons.
```

Perhaps an example as well after this?

Reviewer 2: 1. I think we need "write to the repository _or_ approve changes"
2. I think the use of credentials later in the sentence is still misplaced.

The ACLs systems implement are still based on accounts (users).  User auth to those accounts with their credentials, but the way you ensure they map to a single person is at the account level (since that's where the ACLs are).

Now I'm sure that in some systems account == credential, but those systems may have an even _harder_ (impossible) time enforcing this mapping.  What we really want is that no unique person can appear to be two or more people.

Reviewer 4: > I think we need "write to the repository or approve changes"

you're right.

> The ACLs systems implement are still based on accounts (users). User auth to those accounts with their credentials, but the way you ensure they map to a single person is at the account level (since that's where the ACLs are).

I agree with this where the point of enforcement is a hosted platform that is part of the SCS. I worry this doesn't apply when we're considering other types of systems such as gittuf, that is also listed in the source spec. I do see your point about the word "credentials" not applying well, though.

Reviewer 2: I guess when considering a system like gittuf the goal is still that "No unique person is able to take actions that would otherwise require two or more people."

I don't think talking about the credential actually helps in that case.  As written it's still possible to have two credentials that map to the same person (and in gittuf that probably has to happen in case folks lose a key, are using multiple hardware backed keys, etc?).  Alternatively the advise to implementors is "don't allow a unique person to have more than one credential in the ACL"

Reviewer 2: Oh, but also!  This is _not_ the source requirements doc, this is just _an example_ of a solution for how this can be addressed.

In the future this will link the source reqs doc.

(maybe that needs to be more clear)

Reviewer 4: >Oh, but also! This is not the source requirements doc, this is just an example of a solution for how this can be addressed.

Oh wow, I'd missed that when I paged things in for this PR!

> I guess when considering a system like gittuf the goal is still that "No unique person is able to take actions that would otherwise require two or more people."
>
> I don't think talking about the credential actually helps in that case. As written it's still possible to have two credentials that map to the same person (and in gittuf that probably has to happen in case folks lose a key, are using multiple hardware backed keys, etc?). Alternatively the advise to implementors is "don't allow a unique person to have more than one credential in the ACL"

You're right in that "credentials" doesn't necessarily address it for the gittuf case, which was also my concern for "account". But, given this isn't actually part of the source track, I think we can close this thread. :)

robots.

*Example:* A file within the source repository is automatically generated by a
robot, which is allowed to submit without review.
Adversary compromises the robot and submits a malicious change.
Solution: Require review for such changes.

</details>
<details><summary>Abuse of rule exceptions</summary>

*Threat:* Rule exceptions provide vector for abuse

*Mitigation:* Remove rule exceptions.

*Example:* The intent of a producer is to require two-person review on "all changes except for documentation changes," defined as those only modifying `.md` files.
Adversary submits a malicious executable named `evil.md` and a code review is not required due to the exception.
Technically, the intent of the producer was followed and the produced malicious revision meets all defined policies.

> Solution: Do not allow such exceptions.

Reviewer: Who should not allow such exceptions? The SCS, the org, or someone else?

</details>

<details><summary>Highly-permissioned actor bypasses or disables controls</summary>

*Threat:* Trusted actor with "admin" privileges in a repository submits code by disabling existing controls.

*Mitigation:* All actors must be subject to same controls, whether or not they have
administrator privileges.
Changes to the controls themselves should require their own review process.

*Example 1:* A GitHub repository-level admin pushes a change without review, even though GitHub branch protection is enabled.
Solution: The producer can modify the rule to disallow bypass by administrators, or move the rule to an organization-level ruleset.

*Example 2:* GitHub repository-level admin removes a branch requirement, pushes their change, then re-enables the requirement to cover their tracks.

> Solution: The producer can use higher-permission-level rulesets (such as organization-level) to prevent repository-level tampering.

Reviewer: Have we defined "rulesets" in our terminology?

robots.

*Example:* A file within the source repository is automatically generated by a
robot, which is allowed to submit without review.
Adversary compromises the robot and submits a malicious change.
Solution: Require review for such changes.

</details>
<details><summary>Abuse of rule exceptions</summary>

*Threat:* Rule exceptions provide vector for abuse

*Mitigation:* Remove rule exceptions.

*Example:* The intent of a producer is to require two-person review on "all changes except for documentation changes," defined as those only modifying `.md` files.
Adversary submits a malicious executable named `evil.md` and a code review is not required due to the exception.
Technically, the intent of the producer was followed and the produced malicious revision meets all defined policies.

> Solution: Do not allow such exceptions.

Author: ```suggestion
Solution: The producer adjusts the rules to prohibit such exceptions.
```

</details>

<details><summary>Highly-permissioned actor bypasses or disables controls</summary>

*Threat:* Trusted actor with "admin" privileges in a repository submits code by disabling existing controls.

*Mitigation:* All actors must be subject to same controls, whether or not they have
administrator privileges.
Changes to the controls themselves should require their own review process.

*Example 1:* A GitHub repository-level admin pushes a change without review, even though GitHub branch protection is enabled.
Solution: The producer can modify the rule to disallow bypass by administrators, or move the rule to an organization-level ruleset.

*Example 2:* GitHub repository-level admin removes a branch requirement, pushes their change, then re-enables the requirement to cover their tracks.

> Solution: The producer can use higher-permission-level rulesets (such as organization-level) to prevent repository-level tampering.

Author: ```suggestion
Solution: The producer can configure higher-permission-level rules (such as organization-level GitHub Rulesets) to prevent repository-level tampering.
```

robots.

*Example:* A file within the source repository is automatically generated by a
robot, which is allowed to submit without review.
Adversary compromises the robot and submits a malicious change.

> Solution: Require review for such changes.

Reviewer 2: Optional: 

```suggestion
Solution: Require review for such changes or ensure all changes relevant to the robot's behavior require review.
```

E.g. imagine if the Robot is a GHA that generates code.  If the workflow that runs that code is reviewed, and the generation tool is reviewed, then it could be safe?

Author: (commented on the other thread)

Reviewer 2: I suppose I think there are legitimate use cases for bypassing code review by robots, and I think it is possible to do safely?

I also don't think we have to address that here and now, especially if we don't have a fully fleshed out framework for doing so.

Author: I think eventually we'd need to say something like:
## autobypasses
for some security postures, this practice increases risk because due to threat of account takeover.
For others it's way to reduce risk because security patches are applied faster (or whatever the bot does)

</details>
<details><summary>Abuse of rule exceptions</summary>

*Threat:* Rule exceptions provide vector for abuse

*Mitigation:* Remove rule exceptions.

*Example:* The intent of a producer is to require two-person review on "all changes except for documentation changes," defined as those only modifying `.md` files.

> Adversary submits a malicious executable named `evil.md` and a code review is not required due to the exception.

Reviewer 2: On first reading it wasn't clear to me what the actual problem was here.

```suggestion
Adversary submits a malicious executable masquerading as a documentation file, `help.md`. A code review is not required due to the exception, and a user (or another workflow) is then induced to _execute_ `help.md` and are compromised.
```

*Example:* MyPackage's source repository requires two-person review. Adversary
forks the repo, submits a change in the fork with review from a colluding
colleague (who is not trusted by MyPackage), then merges the change back into
the upstream repo.
Solution: The merge should still require review, even though the fork was reviewed.

</details>

</details>
<details><summary>Commit graph attacks</summary>

*Threat:* Request review for a series of two commits, X and Y, where X is bad and Y is good.
Reviewer thinks they are approving only the final Y state but they are also implicitly approving X.

*Mitigation:* The producer declares that only the final delta is considered approved.

> Intermediate revisions don't count as being reviewed and are not added to the protected context (such as the `main` branch).

Author: ```suggestion
In this configuration, intermediate revisions are not considered to be approved and are not added to the protected context (e.g. the `main` branch).
With git version control systems this is called a "squash" merge strategy.
```

*Example:* Adversary sends a pull request containing malicious commit X and benign commit Y that undoes X.
The produced diff of X + Y contains zero lines of changed code and the reviewer may not notice that X is malicious unless they review each commit in the request.

> If X is allowed to become reachable from the protected branch, the content may become available in secured contexts, such as developer machines and vulnerable to exploits.

Author: ```suggestion
If X is allowed to become reachable from the protected branch, the content may become available in secured environments such as developer machines.
```

*Example:* MyPackage's source repository requires two-person review. Adversary
forks the repo, submits a change in the fork with review from a colluding
colleague (who is not trusted by MyPackage), then proposes the change to
the upstream repo.

> Solution: The proposed change still requires two-person review in the upstream context even though it received two-person review in another context.

Author: I updated the words to be a little more precise. 
In general, I'd expect reviews to be transitive for some situations. but yeah, if your org thinks this is a problem, then this is how you'd fix it.

*Example:* MyPackage's source repository requires two-person review. Adversary
forks the repo, submits a change in the fork with review from a colluding
colleague (who is not trusted by MyPackage), then proposes the change to
the upstream repo.
Solution: The proposed change still requires two-person review in the upstream context even though it received two-person review in another context.

</details>

</details>
<details><summary>Commit graph attacks</summary>

*Threat:* Request review for a series of two commits, X and Y, where X is bad and Y is good.
Reviewer thinks they are approving only the final Y state but they are also implicitly approving X.

*Mitigation:* The producer declares that only the final delta is considered approved.
In this configuration, intermediate revisions are not considered to be approved and are not added to the protected context (e.g. the `main` branch).

> With git version control systems this is called a "squash" merge strategy.

Reviewer 3: While it is likely harder to detect/enforce, another reasonable mitigation is just to review all commits. There could even be automation which flags content from earlier commits that have been removed/overwritten in order to try to detect this type of attack.

Author: that's technically possible, but I would say that getting that right requires multiple squash merge PRs -- you cannot really do it in one shot.

*Example:* MyPackage's source repository requires two-person review. Adversary
forks the repo, submits a change in the fork with review from a colluding
colleague (who is not trusted by MyPackage), then proposes the change to
the upstream repo.
Solution: The proposed change still requires two-person review in the upstream context even though it received two-person review in another context.

</details>

</details>

> <details><summary>Commit graph attacks</summary>

Author: @arewm @TomHennen 

I rewrote this section based on y'all's feedback. 

My goal was to remove the wonky git stuff about squashes and focus on reachable revisions following process.

Reviewer 5: Oh, so one of the things I've done in a previous life was the attestation is actually an attestation of a particular ordering of the commits. so you are attesting to the change not each individual commit so if you roll back a commit it would no longer be valid since you're just attesting to HEAD. with the assumption you're just attesting to old HEAD -> PR/New HEAD

Reviewer 3: I know some folks that hold this to be a hotly debated and contentious topic. It feels right for it to be required at a higher SLSA level because of the higher protections that it guarantees. Some communities preferences may still prefer to not use this approach preventing the SLSA source level that can be achieved.

*Threat:* Two trusted persons collude to author and approve a bad change.

*Mitigation:* This threat is not currently addressed by SLSA. We use "two

> trusted persons" as a proxy for "intent of the software producer".

Author: I do not think we should use "two trusted persons" as a proxy for producer intent. We should probably say "the documented change process for the source" if we must, but I prefer just "the intent".
